### PR TITLE
feat: Setup Slither

### LINKED ISSUE #3: Setup `slither`
What should be done:
1. Fix all slither `info` warnings produced by `slither .`
2. Setup a workflow that runs slither check on any PR and push to the `development` branch (make sure it fails if there are any warnings, [example](https://github.com/ubiquity/ubiquity-dollar/blob/8fd2ec2a608aded67238b17ce62e995e8f3deeb9/.github/workflows/slither-analysis.yml))

### PR DESCRIPTION
**Pull request by gpylypchuk**
Resolves [#3](https://github.com/ubiquity/nft-rewards/issues/3) issue.

### CODE REVIEW

**Comment by gpylypchuk** [comment]
I'm ignoring the Slither warning because the actual contracts of OpenZeppelin are with ```pragma solidity ^0.8.20```

File: src/NftReward.sol [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.13;
+pragma solidity ^0.8.20;
```
[the remark this answers is not in the document]
Comment (0x4007) [reply src/NftReward.sol]: This is really interesting

File: src/NftReward.sol [hunk 2]
```diff
@@ -197,6 +199,7 @@ contract NftReward is Initializable, ERC721Upgradeable, OwnableUpgradeable, Paus
      * @param _newMinter New minter address
      */
     function setMinter(address _newMinter) external onlyOwner {
+        require(_newMinter != address(0), "Minter cannot be zero address");
```
Comment (rndquu) [inline src/NftReward.sol]: Pls add tests for all newly added `require()` statements

**Review by rndquu — CHANGES_REQUESTED** [review]

**Comment by gpylypchuk** [comment]
should I add the merged changes to this PR? (about #4)

**Comment by rndquu** [comment]
> should I add the merged changes to this PR? (about #4)

Yes, you should merge the `development` branch because there are code conflicts which should be resolved

**Comment by gpylypchuk** [comment]
Done @rndquu !

**Review by rndquu — CHANGES_REQUESTED** [review]
1. There is still this warning:
```
NftReward.safeMint(NftReward.MintRequest,bytes) (src/NftReward.sol#145-176) uses timestamp for comparisons
	Dangerous comparisons:
	- require(bool,string)(block.timestamp < _mintRequest.deadline,Signature expired) (src/NftReward.sol#155)
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#block-timestamp
```

I think we can simply exclude the `block-timestamp` detector.

2. There is still this warning:
```
Pragma version^0.8.20 (src/NftReward.sol#2) necessitates a version too recent to be trusted. Consider deploying with 0.8.18.
solc-0.8.24 is not recommended for deployment
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity
```

What if we use solidity v0.8.18?

3. There is still this info warning:
```
Parameter NftReward.invalidateNonce(uint256)._nonceValue (src/NftReward.sol#223) is not in mixedCase
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#conformance-to-solidity-naming-conventions
```

Either exclude this detector either refactor the code.

4. Pls make sure the slither check workflow fails on any warning in the `NftReward.sol` contract

5. Pls update the slither worflow to run only when `*.sol` files are changed

**Comment by gpylypchuk** [comment]
> 1. There is still this warning:
> 
> ```
> NftReward.safeMint(NftReward.MintRequest,bytes) (src/NftReward.sol#145-176) uses timestamp for comparisons
> 	Dangerous comparisons:
> 	- require(bool,string)(block.timestamp < _mintRequest.deadline,Signature expired) (src/NftReward.sol#155)
> Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#block-timestamp
> ```
> 
> I think we can simply exclude the `block-timestamp` detector.
> 
> 2. There is still this warning:
> 
> ```
> Pragma version^0.8.20 (src/NftReward.sol#2) necessitates a version too recent to be trusted. Consider deploying with 0.8.18.
> solc-0.8.24 is not recommended for deployment
> Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity
> ```
> 
> What if we use solidity v0.8.18?
> 
> 3. There is still this info warning:
> 
> ```
> Parameter NftReward.invalidateNonce(uint256)._nonceValue (src/NftReward.sol#223) is not in mixedCase
> Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#conformance-to-solidity-naming-conventions
> ```
> 
> Either exclude this detector either refactor the code.
> 
> 4. Pls make sure the slither check workflow fails on any warning in the `NftReward.sol` contract
> 5. Pls update the slither worflow to run only when `*.sol` files are changed

In 2, I tested before and there is still the warning because of the OZ libraries

**Comment by rndquu** [comment]
> In 2, I tested before and there is still the warning because of the OZ libraries

But there is no warning for `NftReward.sol` in this case, right?

**Comment by gpylypchuk** [comment]
> > In 2, I tested before and there is still the warning because of the OZ libraries
> 
> But there is no warning for `NftReward.sol` in this case, right?

correct!

**Comment by rndquu** [comment]
> > > In 2, I tested before and there is still the warning because of the OZ libraries
> > 
> > 
> > But there is no warning for `NftReward.sol` in this case, right?
> 
> correct!

Then it makes sense to update `NftReward.sol` to use solidity `v0.8.18` since OZ contracts are not part of this project

File: src/NftReward.sol [hunk 3]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.13;
+pragma solidity ^0.8.18;
```
[the remark this answers is not in the document]
Comment (gpylypchuk) [reply src/NftReward.sol:2]: @rndquu what do you think?

[the remark this answers is not in the document]
Comment (rndquu) [reply src/NftReward.sol:2]: Perhaps if https://github.com/crytic/slither-action targets only the https://github.com/ubiquity/nft-rewards/tree/development/src folder the OZ related warnings would be gone?

**Comment by gpylypchuk** [comment]
Should we fail on medium or low? Also I think we should Ignore the OZ libraries 🤔

File: src/NftReward.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (gpylypchuk) [reply src/NftReward.sol:2]: Yes I am thinking that

**Comment by rndquu** [comment]
> Should we fail on medium or low? Also I think we should Ignore the OZ libraries 🤔

Yes, we use this strategy (fall on any slither warning or compiler warning) in our main repo https://github.com/ubiquity/ubiquity-dollar/

Not all of the slither detectors are included in our main repo because it makes sense to be reasonable (for example, `block.timestamp` related warnings can be excluded)

**Comment by gpylypchuk** [comment]
> > Should we fail on medium or low? Also I think we should Ignore the OZ libraries 🤔
> 
> Yes, we use this strategy (fall on any slither warning or compiler warning) in our main repo https://github.com/ubiquity/ubiquity-dollar/
> 
> Not all of the slither detectors are included in our main repo because it makes sense to be reasonable (for example, `block.timestamp` related warnings can be excluded)

yep, I added a detector ignore about OZ libraries & timestamp

**Comment by gpylypchuk** [comment]
@rndquu should I add this: ```"exclude_informational": true``` to ignore this ```solc-0.8.24 is not recommended for deployment
Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity``` ?

**Comment by rndquu** [comment]
> @rndquu should I add this: `"exclude_informational": true` to ignore this `solc-0.8.24 is not recommended for deployment Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity` ?

If we remove the [exclude_informational](https://github.com/ubiquity/nft-rewards/pull/7/files#diff-959713dbf93e19e462fad9058feb6086c9c053d875e23c9ffc637ac12c26ea52R4) config the slither workflow still passes, right?

**Comment by gpylypchuk** [comment]
> > @rndquu should I add this: `"exclude_informational": true` to ignore this `solc-0.8.24 is not recommended for deployment Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity` ?
> 
> If we remove the [exclude_informational](https://github.com/ubiquity/nft-rewards/pull/7/files#diff-959713dbf93e19e462fad9058feb6086c9c053d875e23c9ffc637ac12c26ea52R4) config the slither workflow still passes, right?

No, I think that is because is like a low vuln.
In [this](https://github.com/ubiquity/nft-rewards/pull/7/commits/0db9bf72729b4ff84b1813f7cbd0c0ed338dd11b) commit I removed the ```exclude_informational``` and the log show that failed because of that.

**Comment by rndquu** [comment]
> > > @rndquu should I add this: `"exclude_informational": true` to ignore this `solc-0.8.24 is not recommended for deployment Reference: https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity` ?
> > 
> > 
> > If we remove the [exclude_informational](https://github.com/ubiquity/nft-rewards/pull/7/files#diff-959713dbf93e19e462fad9058feb6086c9c053d875e23c9ffc637ac12c26ea52R4) config the slither workflow still passes, right?
> 
> No, I think that is because is like a low vuln. In [this](https://github.com/ubiquity/nft-rewards/pull/7/commits/0db9bf72729b4ff84b1813f7cbd0c0ed338dd11b) commit I removed the `exclude_informational` and the log show that failed because of that.

I don't understand from the log output, what file causes the `solc-0.8.24 is not recommended for deployment` error?

**Comment by gpylypchuk** [comment]
> exclude_informational

That is because slither only recommends deployment with the version [0.8.18](https://github.com/crytic/slither/wiki/Detector-Documentation#incorrect-versions-of-solidity) and the OZ contracts are in  ^0.8.20 so there is no way to not recognize that error.

**Comment by gpylypchuk** [comment]
I just added a line in ```foundry.toml``` about ```solc_version = "0.8.20"``` so you can select the version you want to compile.

**Comment by rndquu** [comment]
> I just added a line in `foundry.toml` about `solc_version = "0.8.20"` so you can select the version you want to compile.

But that doesn't solve the warning as far as I understand

Overall we shouldn't exclude all low level issues. This is getting out of hand so lets:
1. Remove `exclude_informational: true`
2. Add the `solc-version` detector to exceptions. The only valid case I can think of is the usage of the `PUSH0` instruction which is mainly supported.

**Comment by gpylypchuk** [comment]
> > I just added a line in `foundry.toml` about `solc_version = "0.8.20"` so you can select the version you want to compile.
> 
> But that doesn't solve the warning as far as I understand
> 
> Overall we shouldn't exclude all low level issues. This is getting out of hand so lets:
> 
> 1. Remove `exclude_informational: true`
> 2. Add the `solc-version` detector to exceptions. The only valid case I can think of is the usage of the `PUSH0` instruction which is mainly supported.

done!

**Review by rndquu — APPROVED** [review]

**Review by gitcoindev — APPROVED** [review]
Reviewed, read all comments and approving from my side, too!

**Pull request merged**